pkg/descheduler/strategies/pod_antiaffinity.go

```diff
@@ -0,0 +1,94 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package strategies
+
+import (
+	"github.com/kubernetes-incubator/descheduler/cmd/descheduler/app/options"
+	"github.com/kubernetes-incubator/descheduler/pkg/api"
+	"k8s.io/kubernetes/pkg/api/v1"
+
+	"github.com/golang/glog"
+	"github.com/kubernetes-incubator/descheduler/pkg/descheduler/evictions"
+	podutil "github.com/kubernetes-incubator/descheduler/pkg/descheduler/pod"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	//TODO: Change to client-go instead of generated clientset.
+	"k8s.io/kubernetes/pkg/client/clientset_generated/clientset"
+	priorityutil "k8s.io/kubernetes/plugin/pkg/scheduler/algorithm/priorities/util"
+	"k8s.io/kubernetes/plugin/pkg/scheduler/schedulercache"
+)
+
+// RemovePodsWithAffinityRules with elimination strategy
+func RemovePodsWithAffinityRules(ds *options.DeschedulerServer, strategy api.DeschedulerStrategy, policyGroupVersion string, nodes []*v1.Node) {
+	if !strategy.Enabled {
+		return
+	}
+	removePodsWithAffinityRules(ds.Client, policyGroupVersion, nodes, ds.DryRun)
+}
+
+// removePodsWithAffinityRules evicts pods on the node which are having a pod affinity rules.
+func removePodsWithAffinityRules(client clientset.Interface, policyGroupVersion string, nodes []*v1.Node, dryRun bool) int {
+	podsEvicted := 0
+	for _, node := range nodes {
+		glog.V(1).Infof("\nProcessing node: %#v\n", node.Name)
+		pods, err := podutil.ListPodsOnANode(client, node)
+		if err != nil {
+			return 0
+		}
+		for _, pod := range pods {
+			if checkPodsWithAntiAffinityExist(pod, pods) {
+				success, err := evictions.EvictPod(client, pod, policyGroupVersion, dryRun)
+				if !success {
+					glog.Infof("Error when evicting pod: %#v (%#v)\n", pod.Name, err)
+				} else {
+					podsEvicted++
+					glog.V(1).Infof("Evicted pod: %#v (%#v)\n as because of existing anti-affinity", pod.Name, err)
+				}
+			}
+		}
+	}
+	return podsEvicted
+}
+
+// As of now, ignoring soft or hard anti-affinity rules(?) -
```

Comment:
you are considering hard anti-affinity (RequiredDuringSchedulingIgnoredDuringExecution) and ignorning soft anti-affinty (PeferredDuringSchedulingIgnoredDuringExecution).

Reply:
Ok, what I meant was - which one should I consider? hard or soft or both?

Reply:
hard which you are doing anyway. Just update the comment.